library/L2_HAL/displays/lcd/test/st7066u_test.cpp

```diff
@@ -0,0 +1,343 @@
+// Tests for the St7066u Parallel LCD Driver class.
+#include "L2_HAL/displays/lcd/st7066u.hpp"
+#include "L2_HAL/io/parallel_bus.hpp"
+#include "L4_Testing/testing_frameworks.hpp"
+
+namespace sjsu
+{
+EMIT_ALL_METHODS(St7066u);
+
+TEST_CASE("Testing St7066u Parallel LCD Driver", "[st70668]")
+{
+  Mock<Gpio> mock_rs;  // RS: Register Select
+  Mock<Gpio> mock_rw;  // RW: Read / Write
+  Mock<Gpio> mock_e;   // E   Chip Enable
+  Mock<ParallelBus> mock_data_bus;
+
+  Fake(Method(mock_rs, SetDirection), Method(mock_rs, Set));
+  Fake(Method(mock_rw, SetDirection), Method(mock_rw, Set));
+  Fake(Method(mock_e, SetDirection), Method(mock_e, Set));
+  Fake(Method(mock_data_bus, Initialize),
+       Method(mock_data_bus, SetDirection),
+       Method(mock_data_bus, Write));
+
+  const St7066u kFourBitBusLcd(St7066u::BusMode::kFourBit,
+                               St7066u::DisplayMode::kSingleLine,
+                               St7066u::FontStyle::kFont5x8,
+                               mock_rs.get(),
+                               mock_rw.get(),
+                               mock_e.get(),
+                               mock_data_bus.get());
+  const St7066u kEightBitBusLcd(St7066u::BusMode::kEightBit,
+                                St7066u::DisplayMode::kSingleLine,
+                                St7066u::FontStyle::kFont5x8,
+                                mock_rs.get(),
+                                mock_rw.get(),
+                                mock_e.get(),
+                                mock_data_bus.get());
+
+  SECTION("Initialize")
+  {
+    constexpr St7066u::Command kDefaultConfiguration =
+        St7066u::Command::kDefaultDisplayConfiguration;
+    constexpr St7066u::DisplayMode kDisplayMode =
+        St7066u::DisplayMode::kSingleLine;
+    constexpr St7066u::FontStyle kFontStyle = St7066u::FontStyle::kFont5x8;
+    St7066u::BusMode bus_mode;
+    /// The expected bytes that are sent for initialization.
+    std::array<uint8_t, 6> expected_data;
+    size_t write_count = 0;
+
+    When(Method(mock_data_bus, Write))
+        .AlwaysDo([&expected_data, &write_count](uint32_t data) {
+          INFO("write_count: " << write_count);
+          INFO("data: " << data);
+          CHECK(data == expected_data[write_count]);
+          write_count++;
+        });
+
+    SECTION("With 4-bit Bus")
+    {
+      bus_mode                         = St7066u::BusMode::kFourBit;
+      const uint8_t kConfigurationData = Value(kDefaultConfiguration) |
+                                         Value(bus_mode) | Value(kDisplayMode) |
+                                         Value(kFontStyle);
+      // For 4-bit bus mode, the most significant nibble is sent first followed
+      // by the least significant nibble.
+      expected_data = {
```

Comment:
A little confusing. Maybe put a comment stating that this array gets checked in the `AlwaysDo` statement above.